include/odp/api/spec/traffic_mngr.h

```diff
@@ -189,6 +189,30 @@ extern "C" {
  * tree/hierarchy of nodes.
  */
 
+/** The tm_node_stats_type enumeration lists possible packet or octet
+ * statistics at a tm node.
+ */
+typedef enum odp_tm_node_stats_type_t {
+	/** Packets dropped by this node after scheduling/shaping at this node */
+	ODP_TM_NODE_STATS_PKTS_DROPPED,
+	/** Octets dropped after scheduling/shaping at this node */
+	ODP_TM_NODE_STATS_OCTETS_DROPPED,
+	/** Green packets that are sent through this tm node */
+	ODP_TM_NODE_STATS_GREEN_PKTS,
+	/** Green octets that are sent through this tm node */
+	ODP_TM_NODE_STATS_GREEN_OCTETS,
+	/** Yellow packets that are sent through this tm node */
+	ODP_TM_NODE_STATS_YELLOW_PKTS,
+	/** Yellow octets that are sent through this tm node */
+	ODP_TM_NODE_STATS_YELLOW_OCTETS,
+	/** Red packets that are sent through this tm node */
+	ODP_TM_NODE_STATS_RED_PKTS,
+	/** Red octets that are sent through this tm node */
+	ODP_TM_NODE_STATS_RED_OCTETS,
```

Comment:
Just to check if we have the same understanding: The color referred to here is the odp_packet_color(pkt) of the packet as it entered the TM system and the color never changes within the TM system. Right?

Reply:
We don't currently support pre-colored packets on Tx. And all our shapers are based on RFC2698 like meter were packet color is modified and action of drop or forward to next shaper is decided and hence packet color internally changes as it progresses.

Reply:
It seems that is not according to the API. As far as I can see, the API does not say that odp_packet_color() of a packet can change inside a TM system, and looking at the linux-gen implementation, the color indeed does not change (AFAICS) and e.g. egress VLAN marking is done based on the original, unchanged packet color that the packet had when it entered the TM system.

I am not sure if this makes a lot of sense, but that is how both the API spec and the linux-gen implementation look like now. Maybe API changes are needed.

Reply:
We also currently don't support any marking feature in our implementation such as VLAN, IP ECN or IP DSCP. 
One thing is spec doesn't and might not be able to say how single rate or dual rate shaping is done in a particular HW. 
And RFC2698 is one method of doing so and then coloring a packet based on the result.

However unless packet marking is supported which updates the drop precedence based on final color, I guess coloring packets is not of much use.

>egress VLAN marking is done based on the original, unchanged packet color that the packet had when it entered the TM system.

Below is the text from doc/users-guide/users-guide-tm.adoc. which talks about final packet color.
I couldn't find anything about odp_packet_color() implication in traffic_mngr.h

> ==== Packet Marking
> 
> The Packet Marking API is used to mark the packet based upon the final packet
> color assigned to it when it reaches the egress node.
> This is an optional feature and if available on the platform is used to reflect
> the packet color on IPv4/IPv6 DiffServ filed in accordance with https://www.ietf.org/rfc/rfc2474.txt[RFC 2474].
...
...